### PR TITLE
Format code-blocks (RST-only, MVP)

### PR DESCRIPTION
**Pull request by nathan-contino-mongo**
- create RST codeblocks from code snippets
- support some common languages (coincidentally, those supported by the MongoDB Realm product)
- emphasize lines supported

### CODE REVIEW

File: src/bluehawk/commands/EmphasizeCommand.ts [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+export const EmphasizeCommand = makeBlockOrLineCommand<NoAttributes>({
+  attributesSchema: NoAttributesSchema,
+
+  process({ commandNode, parseResult }) {
+    const { source } = parseResult;
+    const { text } = source;
+
+    // Strip tags
+    removeMetaRange(text, commandNode);
+    commandNode.attributes = {};
```
Comment (cbush) [inline src/bluehawk/commands/EmphasizeCommand.ts:18]: Actually, you'll want to set the file attributes on the `source` object, not the command node.

Also, be sure not to clobber any existing attributes (as is happening here - what if commandNodes already had an attributes list?)

File: src/cli/commands/snip.test.ts [hunk 2]
```diff
@@ -0,0 +1,48 @@
+import { Stats } from "fs";
+import { System } from "../../bluehawk";
+import { snip } from "./snip";
+
+describe("snip", () => {
+  it("snips", async (done) => {
+    System.fs = {
```
Comment (cbush) [inline src/cli/commands/snip.test.ts:7]: This was already updated on the latest main branch, so be sure to fetch and rebase.

File: src/cli/options.ts [hunk 3]
```diff
@@ -54,3 +54,14 @@ export function withStateOption<T>(
     once: true,
   });
 }
+
+export function withGenerateFormattedCodeSnippetsOption<T>(
+  yargs: Argv<T>
+): Argv<T & { generateFormattedCodeSnippets?: string }> {
+  return option(yargs, "format", {
```
Comment (cbush) [inline src/cli/options.ts:61]: Which format? Sphynx style rST is not the only way this could be done. That said, I like the idea of potentially accepting a generic "format" flag and having "sphynx-literalinclude" be one of possible values for it.

File: src/cli/commands/snip.ts [hunk 4]
```diff
@@ -73,6 +70,42 @@ const handler = async ({
...
+
+    // TODO: need to make sure that "generate formatted snippets" is turned on here
+    const targetPath = path.join(destination, source.basename);
+
+    const rstHeader = ".. code-block:: ";
+    const rstEmphasizeModifier = ":emphasize-lines: ";
+    const range = source.attributes["emphasize"]["range"];
+
+    const formattedCodeblock =
+      rstHeader +
```
Comment (cbush) [inline src/cli/commands/snip.ts:91]: Avoid using `+` to concatenate strings - instead either put each line in an array then `.join("\n")` or use template strings (backticks) with `${}` to interpolate values.

File: src/cli/commands/snip.ts [hunk 5]
```diff
@@ -73,6 +70,42 @@ const handler = async ({
...
+  bluehawk.subscribe(async (result: ParseResult) => {
+    const { source } = result;
+    if (
+      source.attributes["snippet"] === undefined ||
+      source.attributes["emphasize"] === undefined
+    ) {
+      return;
+    }
+
+    // TODO: need to make sure that "generate formatted snippets" is turned on here
```
Comment (cbush) [inline src/cli/commands/snip.ts:83]: Why not check before adding this listener?

File: src/cli/options.ts (hunk 3 above)
Comment (nathan-contino-mongo) [reply src/cli/options.ts:61]: That's the plan, that you'll feed in "--format=rst" in the first pass and we can always add more in the future.

File: src/bluehawk/commands/EmphasizeCommand.ts [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+  };
+  end: {
+    line: number;
+    column: number;
+  };
+}
+
+export const EmphasizeCommand = makeBlockOrLineCommand<NoAttributes>({
+  name: "emphasize",
+  description: "highlight line(s) in formatted output",
```
Comment (cbush) [inline src/bluehawk/commands/EmphasizeCommand.ts:23]: Maybe we can be more precise/helpful here? (Mentioning CLI from the library should be ok since it's just an example)
```suggestion
  description: "identify line(s) to highlight (see `bluehawk snip --format` command)",
```

File: src/bluehawk/getBluehawk.ts [hunk 7]
```diff
@@ -19,6 +20,8 @@ export const getBluehawk = async (
   pluginPaths?: string | string[]
 ): Promise<Bluehawk> => {
   if (bluehawk === undefined) {
+    bluehawk = new Bluehawk();
```
Comment (cbush) [inline src/bluehawk/getBluehawk.ts:23]: Why this change? Bad merge? The bluehawk is initalized below.

File: src/cli/commands/snip.ts [hunk 8]
```diff
@@ -42,7 +42,7 @@ export const doRst = async (
     [".kt", "kotlin"],
     [".java", "java"],
     [".gradle", "groovy"],
-    [".objc", "objective-c"],
+    [".m", "objectivec"],
```
Comment (nathan-contino-mongo) [inline src/cli/commands/snip.ts:45]: nice!

File: src/cli/commands/snip.ts [hunk 9]
```diff
@@ -19,15 +20,72 @@ interface SnipArgs extends MainArgs {
   destination: string;
   state?: string;
   ignore?: string | string[];
+  format?: string;
```
Comment (cbush) [inline src/cli/commands/snip.ts:23]: It would be nice to use an enum here: `format?: "sphynx-rst"`

File: src/cli/commands/snip.ts [hunk 10]
```diff
@@ -19,15 +20,72 @@ interface SnipArgs extends MainArgs {
...
+): Promise<string | undefined> => {
+  const { source } = result;
+  if (
+    source.attributes["snippet"] === undefined ||
+    source.attributes["emphasize"] === undefined
+  ) {
+    return undefined;
+  }
+
+  // nasty hack to cover the suffixes/rst languages we use most often with Realm.
```
Comment (cbush) [inline src/cli/commands/snip.ts:37]: Let's not bring Realm into this!
```suggestion
  // nasty hack to cover the suffixes/rst languages we use most often
```

File: src/cli/commands/snip.ts [hunk 11]
```diff
@@ -19,15 +20,72 @@ interface SnipArgs extends MainArgs {
...
+  ]);
+  const rstLanguage = rstLanguageMap.has(source.language)
+    ? rstLanguageMap.get(source.language)
+    : "text";
+
+  const rstHeader = ".. code-block::";
+  const rstEmphasizeModifier = ":emphasize-lines:";
+
+  const rstEmphasizeRanges: { start: number; end: number }[] = [];
+  for (const range of source.attributes["emphasize"]["ranges"]) {
```
Comment (cbush) [inline src/cli/commands/snip.ts:58]: Really, we should have `export interface EmphasizeSourceAttributes` in EmphasizeCommand.ts and cast `source.attributes["emphasize"]` to that (or undefined) everywhere we want to use it. That allows you to use the attributes below with a bit more type safety and autocomplete support. That should be the pattern we follow (recognizing that we now use the word "attributes" in a few different contexts already...)

File: src/cli/commands/snip.ts [hunk 12]
```diff
@@ -19,15 +20,72 @@ interface SnipArgs extends MainArgs {
...
+    .join(", ");
+
+  const formattedCodeblock = [
+    `${rstHeader} ${rstLanguage}`,
+    `   ${rstEmphasizeModifier} ${rstFormattedRanges}`,
+    "", // empty line required between rst codeblock declaration and content
+    source.text
+      .toString()
+      .split(/\r\n|\r|\n/)
+      .map((line) => (line === "" ? line : `   ${line}`))
```
Comment (cbush) [inline src/cli/commands/snip.ts:81]:
```suggestion
      .map((line) => (line === "" ? line : `   ${line}`)) // indent each line 3 spaces
```

Comment (cbush) [reply src/cli/commands/snip.ts:81]: It's a good practice to let people know wtf you're trying to do.

File: src/cli/commands/snip.ts [hunk 13]
```diff
@@ -72,6 +130,29 @@ const handler = async ({
     }
   });
 
+  if (format !== undefined) {
```
Comment (cbush) [inline src/cli/commands/snip.ts:133]: This should check specifically which formatter the user wanted.
```suggestion
  if (format === "sphynx-rst") {
```

Comment (cbush) [reply src/cli/commands/snip.ts:133]: You could also report if the format string was not accepted.

**Review by cbush — APPROVED** [review]
Looks good. A few comments here but nothing blocking. Great job!

File: src/cli/commands/snip.ts (hunk 11 above)
Comment (nathan-contino-mongo) [reply src/cli/commands/snip.ts]: Implemented this interface, and now using it (without the `| undefined`, by sheer personal preference).
